posts/2020-05-15-five-years-of-rust.md

```diff
@@ -0,0 +1,497 @@
+with a module. In 1.20 we stabilised associating constants on struct, enums,
+and importantly traits. Making it easier to add rich sets of preset values for
+types in your API, such as common IP addresses or interesting numbers.
+
+#### 2018
+
+**[1.24] — Incremental Compilation** Before 1.24 when you made a change in your
+library rustc would have to re-compile all of the code. Now rustc is a lot
+smarter about caching as much as possible and only needing to re-generate
+what's needed.
+
+**[1.26] — Existential Types** The addition of `impl Trait` gives you expressive
+dynamic APIs with the benefits and performance of static dispatch.
+
+**[1.28] — Global Allocators** Previously you were restricted to using the
+allocator that rust provided. With the global allocator API you can now
+customise your allocator to one that suits your needs. This was an important
+step in enabling the creation of the `alloc` library, another subset of the
+standard library containing only the parts of std that need an allocator like
+`Vec` or `String`. Now it's easier than ever to use even more parts of the
+standard library on a variety of systems.
+
+**[1.31] — 2018 edition** The release of the 2018 edition was easily our biggest
+release since 1.0, adding a collection of syntax changes and improvements to
+writing Rust written in a completely backwards compatible fashion, allowing
+libraries built with different editions to seamlessly work together.
+
+- **Non-Lexical Lifetimes** A huge improvement to Rust's borrow checker,
+  allowing it to accept more verifiable safe code.
+- **Module System Improvements** Large UX improvements to how we define and
+  use modules.
+- **Const Functions** Const functions allow you to run and evaluate Rust code
+  at compile time.
+- **Rustfmt** A new code formatting tool built specifically for Rust.
+- **Clippy** Rust's linter for catching common mistakes. Clippy makes it a lot
+  easier to make sure that your code is not only safe but correct.
+- **Rustfix** With all the syntax changes, we knew we wanted to provide the
+  tooling to make the transition as easy as possible. Now when changes are
+  required to Rust's syntax they're just a `cargo fix` away from being resolved.
+
+#### 2019
+
+**[1.34] — Alternative Crate Registries** As Rust is used more and more in
+production, there is a greater need to be able to host and use your projects
+in non-public spaces, while cargo has always allowed remote git dependencies,
+with Alternative Registries your organisation can easily build and share your
+own registry of crates that can be used in your projects like they were
+on crates.io.
+
+**[1.39] — Async/Await** The stabilisation of the async/await keywords for
+handling Futures was one of the major milestones to making async programming
+in Rust a first class citizen. Even just six months after its release the
+async programming has blossomed into a diverse and performant ecosystem.
+
```

Comment:
Has nothing happened this year? :-( It would be cool to have at least one new thing for 2020

Reply:
I feel like we stabilized slice patterns, maybe some other pattern additions? We've also seen a lot of progress around const evaluation? But maybe those things didn't hit stable releases yet.

Reply:
There's been three major releases this year [1.41], [1.42], and [1.43]. My thought process when deciding which changes to include was "Would this be interesting to someone who has never written Rust? Would reading about the feature make someone interested in trying Rust?". These past three releases haven't really had that. The main additions are in 1.42 where as @nikomatsakis said we stabilised slice patterns and added improved unwrap error line reporting. My initial thought was that those two features were too small (relatively), if other people feel differently on what to add from these releases I'm happy to add something.

[1.41]: https://blog.rust-lang.org/2020/01/30/Rust-1.41.0.html
[1.42]: https://blog.rust-lang.org/2020/03/12/Rust-1.42.html
[1.43]: https://blog.rust-lang.org/2020/04/23/Rust-1.43.0.html

Reply:
They probably are too small, but if that's it, then I think I would put them in just so we have *something* for 2020

Reply:
I have kind of mixed feelings here. I think it's kind of nice that we haven't had any really major changes this year, especially given the "rate of change of Rust" discussions. But on the other hand, "here's some good news from 2020" is a distressingly rare sentence, and maybe we can provide some such good news. Slice patterns really are pretty cool!